304 find a way to fit the request id into a stellar text memo

TorstenStueber opened the pull request.

ebma commented:
When we had the discussion about the new lookup being `(textmemo -> issue id)` and `(issue id -> issue)` I somehow assumed that this was about the logic in the vault client and not necessarily in the pallet itself. 
Correct me if I'm wrong because maybe I'm missing something but I imagined something simpler where we don't need to change all the extrinsics so much:
- We add a new field, let's call it `text_memo` to the [IssueRequest](https://github.com/pendulum-chain/spacewalk/blob/39b8a257bb178ef1e0823b955c1470b2e95b9ab4/primitives/src/lib.rs#L148) struct (and also to `RedeemRequest` and `ReplaceRequest`)
- Then, when the new issue/redeem/replace is requested, inside the extrinsic we derive the expected memo with your implemented derivation method and put it into the struct we store on chain eg [here](https://github.com/pendulum-chain/spacewalk/blob/39b8a257bb178ef1e0823b955c1470b2e95b9ab4/pallets/issue/src/lib.rs#LL416C22-L416C22).

Now, the vault client somehow needs to know which Stellar transaction text memos it has to match to what Issue/Redeem/Replace request. You implemented that logic already in the vault client, ie we can query all open requests and then derive the to-be-expected memo and store them locally as well. Since we have this `(textmemo -> issue id)` + `(issue id -> issue)` lookup chain available in the vault client, the vault will know the actual ID of the request that it has to fulfill and we don't need to touch the signatures of any of the extrinsics and we also don't need to have an extra on-chain storage item. We just need to add the extra `text_memo` field to the structs so that this data is stored as part of the struct on chain.

Wouldn't this already work or am I missing something @TorstenStueber ?

ebma commented:
@TorstenStueber sounds good. 
And you're right we don't even have to store the memo in the struct, as long as our derivation is 'simple' enough so that any other client software can do that. If we made the derivation more complex it might make sense to store it, but that does not apply for now (and maybe never will) so let's assume we stick with that and don't need to store it on-chain.

TorstenStueber commented:
@ebma I have reverted the extrinsic but now I remember the actual reason why I changed the extrinsic in the first place. The problem was with the task [`execute_issue`](https://github.com/pendulum-chain/spacewalk/blob/39b8a257bb178ef1e0823b955c1470b2e95b9ab4/clients/vault/src/issue.rs#L289-L358).

This task will call the extrinsic `execute_issue` for new Stellar issue transactions. I did not expect that at the point in time this transaction is processed the vault would also already know about the issue.

But I did not read all of the code and just discovered that only transactions are processed that pass the [`is_relevant` filter function](https://github.com/pendulum-chain/spacewalk/blob/39b8a257bb178ef1e0823b955c1470b2e95b9ab4/clients/vault/src/issue.rs#L374-L407). This filter function also explains why my tests are failing – it reads the transaction memo not from the transaction but directly from the transaction response and I did not change the code.

But is this not an actual problem in the current design? A user requests an issue from the pallet (probably directly from some web interface, e.g., the Portal). The vault will listen to a) Stellar transactions on the Stellar ledger and b) for issue requests on the Spacewalk host chain. Whenever it observes a Stellar transaction, it will look up whether it already observed an issue request – [otherwise it will just drop this transaction and never consider again](https://github.com/pendulum-chain/spacewalk/blob/39b8a257bb178ef1e0823b955c1470b2e95b9ab4/clients/wallet/src/horizon.rs#L377-L386) (currently only after a restart as the vault scans all transactions again from the beginning).

But isn't there a race condition here? We must not assume that the vault client observes a) and b) in the expected order. Therefore we should either:
* Keep transactions that transfer to the vault until we observe an according issue request and process them then (I think this is dangerous as information can be lost)
* Forward transactions directly to the `execute_issue` extrinsic without using any knowledge about pending issue requests (but this would mean that the vault would try to replay all historic `execute_issue` extrinsics after restart; not an actual issue as the pallet knows to ignore them but quite wasteful).

ebma commented:
> But isn't there a race condition here? We must not assume that the vault client observes a) and b) in the expected order. 
You could be right and we might have to improve this logic but I think the only problematic scenario is that the connection of our vault client to an RPC node is slower than the connection of a user to an RPC node. 
A user always first has to request a new issue/redeem. He would most likely do that in the portal UI, which shows a loading indicator until the block is _finalized_. Only after it is finalized the user is told to create a Stellar transaction. Now, the vault also listens to events in [blocks that are finalized](https://github.com/pendulum-chain/spacewalk/blob/c57d7ec0b5a6cb6c837e9f6ae1e908f1f626a2a0/clients/runtime/src/rpc.rs#L341). Thus, with _t_ being the time a user needs to create and submit a transaction on Stellar, the vault has _t_ + 6 seconds (Stellar ledger close) to receive the new issue request and note its ID. The slower the RPC connection of the vault client to the node compared to the user <-> node, the worse this gets obviously. But generally, this should not be a problem.

Since we can't rely on things "generally not being a problem", looking at your two options:

> Forward transactions directly to the execute_issue extrinsic without using any knowledge about pending issue requests (but this would mean that the vault would try to replay all historic execute_issue extrinsics after restart; not an actual problem as the pallet knows to ignore them but quite wasteful).

I think we should not do that, as it would be super costly for the vault account as he always has to pay gas fees, also for failing extrinsics. This would also cause some unnecessarily high load on our collator nodes. 

> Keep transactions that transfer to the vault in memory until we observe an according issue request and process them then (this can be gamed and some attacker could just spam transactions which could make the vault use a lot of memory eventually – but nevertheless in this case the vault would already have the problem that after a restart it would take a long time to process all transactions, it can currently only fetch 200 (DEFAULT_PAGE_SIZE) per 5 seconds (POLL_INTERVAL)).

As you already pointed out the vault scans for all transactions again once it restarts. I think to mitigate the problem we face here, instead of keeping everything in memory we can try to rely on the periodic restart. 
I already thought about improving the flow here a bit, because as you said, currently it only fetches the latest 200 transactions. That's why I created ticket #308. With this, the vault would always check all transactions that occurred and it would always be able to find 'lost'/missed transactions. 
Once the network grows and our vault has to handle more and more transactions, this initial fetch will take longer and longer though. I was thinking that maybe we could implement some mechanism, that stores the latest 'fully-processed' paging token, ie it stores the paging token of the transaction up until which all previous transactions were completely processed and upon a periodic restart it would start from that paging token. Maybe this would open another attack window though, because we don't have an easy way to identify which transactions are not and will never be interesting. 
Another idea would be to start fetching the latest transactions from top-to-bottom (ie. fetch the latest first) and then stop fetching for more transactions once transactions are received that are older than the current issue/redeem period ie. older than 24 hours. We can be sure that these can't be used anymore.

TorstenStueber commented:
I agree that your timing assumption is usually true.

> Since we can't rely on things "generally not being a problem", looking at your two options:

I think it is clear to you as well but I just want to reiterate my argument: Spacewalk needs to be 100% watertight conceptually. We must not take assumptions about timing. Every timing assumption will eventually fail for all kinds of reasons (there could be a networking problem that affects it connection to the Substrate host chain but not to Stellar horizon).

For the sake of making progress, I would continue here using the timing assumption and then we can discuss a more watertight solution elsewhere.

TorstenStueber commented on pallets/stellar-relay/Cargo.toml:
```diff
@@ -31,7 +31,7 @@ substrate-stellar-sdk = { git = "https://github.com/pendulum-chain/substrate-ste
 sha2 = { version = "0.10.6", default-features = false }
 
-primitives = { package = "spacewalk-primitives", path = "../../primitives", default-features = false, optional = true }
+primitives = { package = "spacewalk-primitives", path = "../../primitives", default-features = false }
```
This now needs to be an unconditional dependency.

ebma commented on pallets/stellar-relay/src/lib.rs:
```diff
@@ -689,20 +698,20 @@ pub mod pallet {
-		let memo_hash = match transaction_envelope {
-			TransactionEnvelope::EnvelopeTypeTxV0(tx_env) => match tx_env.tx.memo {
-				Memo::MemoHash(hash) => Some(hash),
+	fn get_text_hash_from_tx_env(transaction_envelope: &TransactionEnvelope) -> Option<&Vec<u8>> {
```
Maybe rename to `get_text_memo_from_tx_env`

TorstenStueber replied:
I changed the name and moved the function to the `primitives` package as this was duplicated code.

ebma reviewed (commented):
I added some nitpicky comments. Looks good to me overall. I was a bit confused because you now only modified the text memo stuff for issue requests and not the other two, but there is no good reason to change it for all three besides making it consistent. Probably it's more secure if we just use this for issue requests so fine for me.

TorstenStueber commented on clients/wallet/src/stellar_wallet.rs:
```diff
@@ -83,7 +85,7 @@ impl StellarWallet {
 		asset: Asset,
 		stroop_amount: i64,
-		memo_hash: Hash,
+		memo_hash: [u8; 32],
```
Okay. Done.

TorstenStueber commented on clients/vault/tests/vault_integration_tests.rs:
```diff
@@ -744,7 +755,7 @@ async fn test_issue_overpayment_succeeds() {
 		let stellar_asset =
 			primitives::AssetConversion::lookup(issue.asset).expect("Asset not found");
-		let memo_hash = issue.issue_id.0;
+		let memo_hash = issue.issue_id;
```
I removed these variables and renamed the variable in `send_payment_to_address` from `memo_hash` to `request_id`. Same for the other comments below.

TorstenStueber commented on clients/vault/src/issue.rs:
```diff
@@ -370,19 +416,26 @@ impl IssueFilter {
+			Some(issue_id) => issue_id,
 			None => return false,
-			Some(memo) => H256::from_slice(memo),
 		};
```
The `unwrap()` feels slightly dirty. I implemented it a little different now.

TorstenStueber commented on clients/vault/src/issue.rs:
```diff
@@ -370,19 +416,26 @@ impl IssueFilter {
+		issue_requests: &IssueRequestsMap,
+		issue_memos: &IssueMemoMap,
+	) -> bool {
 		// try to convert the contained memo_hash (if any) to a h256
```
Okay. Done.

TorstenStueber commented on clients/vault/src/issue.rs:
```diff
@@ -314,7 +350,13 @@ pub async fn execute_issue(
-	if let Some(issue_id) = get_issue_id_from_tx_env(&tx_env) {
+	let (issue_id, issue_memo) = match get_issue_memo_from_tx_env(&tx_env) {
+		Some(issue_memo) =>
+			(issue_memos.read().await.get(issue_memo).map(|hash| hash.clone()), Some(issue_memo)),
```
Okay. Done.

TorstenStueber commented on clients/vault/src/issue.rs:
```diff
@@ -145,18 +173,22 @@ pub async fn listen_for_executed_issues(
-		_ => return None,
+/// Returns issue memo of the given TransactionEnvelope, or None if the transaction does
+/// not contain an issue memo
+fn get_issue_memo_from_tx_env(transaction_envelope: &TransactionEnvelope) -> Option<&Vec<u8>> {
```
Okay. Done.

TorstenStueber commented:
@ebma I resolved all your comments. It was basically just a big variable renaming to some more reasonable names.
